zapcore/write_syncer_test.go

```diff
@@ -65,6 +71,80 @@ func TestAddSyncWriter(t *testing.T) {
 	assert.NoError(t, ws.Sync(), "Unexpected error calling a no-op Sync method.")
 }
 
+func TestBufferWriter(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	// If we pass a plain io.Writer, make sure that we still get a WriteSyncer
+	// with a no-op Sync.
+	t.Run("sync", func(t *testing.T) {
+		buf := &bytes.Buffer{}
+		ws, close := Buffer(AddSync(buf), 0, 0)
+		defer close()
+		requireWriteWorks(t, ws)
+		assert.Empty(t, buf.String(), "Unexpected log calling a no-op Write method.")
+		assert.NoError(t, ws.Sync(), "Unexpected error calling a no-op Sync method.")
+		assert.Equal(t, "foo", buf.String(), "Unexpected log string")
+	})
+
+	t.Run("close", func(t *testing.T) {
+		buf := &bytes.Buffer{}
+		ws, close := Buffer(AddSync(buf), 0, 0)
+		requireWriteWorks(t, ws)
+		assert.Empty(t, buf.String(), "Unexpected log calling a no-op Write method.")
+		close()
+		assert.Equal(t, "foo", buf.String(), "Unexpected log string")
+	})
+
+	t.Run("wrap twice", func(t *testing.T) {
+		buf := &bytes.Buffer{}
+		bufsync, close1 := Buffer(AddSync(buf), 0, 0)
+		ws, close2 := Buffer(bufsync, 0, 0)
+		requireWriteWorks(t, ws)
+		assert.Equal(t, "", buf.String(), "Unexpected log calling a no-op Write method.")
+		close2()
+		close1()
+		assert.Equal(t, "foo", buf.String(), "Unexpected log string")
+	})
+
+	t.Run("small buffer", func(t *testing.T) {
+		buf := &bytes.Buffer{}
+		ws, close := Buffer(AddSync(buf), 5, 0)
+		defer close()
+		requireWriteWorks(t, ws)
+		assert.Equal(t, "", buf.String(), "Unexpected log calling a no-op Write method.")
+		requireWriteWorks(t, ws)
+		assert.Equal(t, "foo", buf.String(), "Unexpected log string")
+	})
+
+	t.Run("flush error", func(t *testing.T) {
+		ws, close := Buffer(AddSync(&errorWriter{}), 4, time.Nanosecond)
+		n, err := ws.Write([]byte("foo"))
+		require.NoError(t, err, "Unexpected error writing to WriteSyncer.")
+		require.Equal(t, 3, n, "Wrote an unexpected number of bytes.")
+		ws.Write([]byte("foo"))
+		assert.Error(t, close(), "Expected close to fail.")
+	})
+
+	t.Run("flush timer", func(t *testing.T) {
+		buf := &bytes.Buffer{}
+		ws, close := Buffer(AddSync(buf), 6, time.Microsecond)
+		defer close()
+		requireWriteWorks(t, ws)
+		ztest.Sleep(10 * time.Millisecond)
+		bws := ws.(*bufferWriterSyncer)
+		bws.Lock()
+		assert.Equal(t, "foo", buf.String(), "Unexpected log string")
+		bws.Unlock()
+
+		// flush twice to validate loop logic
+		requireWriteWorks(t, ws)
+		ztest.Sleep(10 * time.Millisecond)
+		bws.Lock()
+		assert.Equal(t, "foofoo", buf.String(), "Unexpected log string")
+		bws.Unlock()
```

Comment:
why is the lock necessary here? there should be nothing to flush, so the `buf` should not be mutated while we read this (and if it does get mutated, that's likely a real race issue to investigate)?

Reply:
Because we write the buf in goroutine 20, and read the buf in goroutine 19, maybe this race log says all of it.
And it should only reach race condition in this test case, because we need to check what the buf contains by `buf.String()`.
```
Read at 0x00c000117560 by goroutine 19:
  bytes.(*Buffer).String()
      /home/liqi/workspace/go/src/bytes/buffer.go:65 +0x411
  go.uber.org/zap/zapcore.TestBufferWriter.func6()
      /home/liqi/workspace/zap/zapcore/write_syncer_test.go:134 +0x1a4
  testing.tRunner()
      /home/liqi/workspace/go/src/testing/testing.go:1127 +0x202

Previous write at 0x00c000117560 by goroutine 20:
  bytes.(*Buffer).grow()
      /home/liqi/workspace/go/src/bytes/buffer.go:128 +0x484
  bytes.(*Buffer).Write()
      /home/liqi/workspace/go/src/bytes/buffer.go:172 +0x184
  go.uber.org/zap/zapcore.(*writerWrapper).Write()
      <autogenerated>:1 +0x87
  bufio.(*Writer).Flush()
      /home/liqi/workspace/go/src/bufio/bufio.go:607 +0x13c
  go.uber.org/zap/zapcore.(*bufferWriterSyncer).Sync()
      /home/liqi/workspace/zap/zapcore/write_syncer.go:161 +0xb8
  go.uber.org/zap/zapcore.(*bufferWriterSyncer).flushLoop()
      /home/liqi/workspace/zap/zapcore/write_syncer.go:173 +0x87

Goroutine 19 (running) created at:
  testing.(*T).Run()
      /home/liqi/workspace/go/src/testing/testing.go:1178 +0x796
  go.uber.org/zap/zapcore.TestBufferWriter()
      /home/liqi/workspace/zap/zapcore/write_syncer_test.go:128 +0x1cc
  testing.tRunner()
      /home/liqi/workspace/go/src/testing/testing.go:1127 +0x202

Goroutine 20 (running) created at:
  go.uber.org/zap/zapcore.Buffer()
      /home/liqi/workspace/zap/zapcore/write_syncer.go:133 +0x32b
  go.uber.org/zap/zapcore.TestBufferWriter.func6()
      /home/liqi/workspace/zap/zapcore/write_syncer_test.go:130 +0xde
  testing.tRunner()
      /home/liqi/workspace/go/src/testing/testing.go:1127 +0x202
```